server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -4181,10 +4183,25 @@ public void waitForSegmentGeneration(long segmentGeneration, ActionListener<Long
         waitForPrimaryTermAndGeneration(getOperationPrimaryTerm(), segmentGeneration, listener);
     }
 
+    private final SubscribableListener<Void> waitForPrimaryTermAndGenerationListeners = new SubscribableListener<Void>();
+
+    private void checkAndCallWaitForPrimaryTermAndGenerationListeners() {
+        if (state == IndexShardState.STARTED || state == IndexShardState.CLOSED) {
```

Comment:
Again a question for my learning, since what we need is an non-null engine, why not checking `getEngineOrNull` return non-null instead?

Also I am not quite sure the usefulness of adding the primaryTermAndGeneration listener when the shard is closed. It seems that returning error is more plausible? But I guess it is mostly to maintain the existing behaviour?